File: Algorithms/p343_Integer_Break/p343_Integer_Break_1.py
class Solution(object):
    def integerBreak(self, n):
        """
        :type n: int
        :rtype: int
        """
        f = [0]*(n+1)
        g = [0]*(n+1)
        g[1] = 1
        for i in range(2,n+1):
            for j in range(1,i):
                v = j*g[i-j]
                if v>f[i]:
                    f[i] = v
            g[i] = max(i,f[i])
        return f[n]

File: Algorithms/p343_Integer_Break/p343_Integer_Break_1.cpp
class Solution {
public:
    int integerBreak(int n) {
        vector<int> f(n+1);
        vector<int> g(n+1);
        g[1] = 1;
        for(int i=2;i<n+1;i++){
            int max = 0;
            for(int j=1;j<i;j++){
                int v = j*g[i-j];
                if(v>max){
                    max = v;
                }
            }
            f[i] = max;
            g[i] = max>i ? max : i;
        }
        return f[n];
    }
};
